### PR TITLE
IDE-1032 Add ability to override the ECL Watch location

### PR DESCRIPTION
**Pull request by dehilsterlexis**
Signed-off-by: David de Hilster <david.dehilster@lexisnexisrisk.com>

### CODE REVIEW

**Comment by dehilsterlexis** [comment]
@GordonSmith for review

File: eclide/ResultView.cpp [hunk 1]
```diff
@@ -100,13 +104,27 @@ const TCHAR * GetDesdlEclWatchURL(const std::_tstring & desdlID, const std::_tst
...
     url = str.c_str();
+    url = OverRideEclWatchURL(url);
     url += desdlID.c_str();
     url += _T(".");
     url += desdlVersion.c_str();
     url = url.Trim();
     return url;
 }
 
+const TCHAR* OverRideEclWatchURL(CString& url)
```
Comment (GordonSmith) [inline eclide/ResultView.cpp:115]: Should be `const CString& url`

File: eclide/ResultView.cpp [hunk 2]
```diff
@@ -100,13 +104,27 @@ const TCHAR * GetDesdlEclWatchURL(const std::_tstring & desdlID, const std::_tst
...
+{
+    bool advancedFlag = (bool)GetIConfig(QUERYBUILDER_CFG)->Get(GLOBAL_SERVER_ADVANCED);
+    CString returnUrl = url;
+    if (advancedFlag > 0) {
```
Comment (GordonSmith) [inline eclide/ResultView.cpp:119]: Why `> 0`?

File: eclide/ResultView.cpp [hunk 3]
```diff
@@ -100,13 +104,27 @@ const TCHAR * GetDesdlEclWatchURL(const std::_tstring & desdlID, const std::_tst
...
+        const TCHAR* findStr = _T(":8010");
```
Comment (GordonSmith) [inline eclide/ResultView.cpp:120]: Might not contain 8010 (esp in advanced mode)

File: eclide/ResultView.cpp [hunk 4]
```diff
@@ -49,6 +49,7 @@ const TCHAR * GetFramedWorkUnitEclWatchURL(Dali::IWorkunit *wu, CString &url)
     std::_tstring str = static_cast<const TCHAR * >(CString(GetIConfig(QUERYBUILDER_CFG)->Get(GLOBAL_SERVER_WORKUNIT)));
     boost::algorithm::ireplace_first(str, _T("/WsWorkunits"), _T("/esp/files/stub.htm?Wuid="));
     url = str.c_str();
+    url = OverRideEclWatchURL(url);
```
Comment (GordonSmith) [inline eclide/ResultView.cpp:52]: Instead of this, why not change line 49 to use GLOBAL_ECLWATCH?

File: eclide/ResultView.cpp [hunk 5]
```diff
@@ -61,6 +62,7 @@ const TCHAR * GetWorkUnitEclWatchURL(Dali::IWorkunit *wu, CString &url)
     std::_tstring str = static_cast<const TCHAR * >(CString(GetIConfig(QUERYBUILDER_CFG)->Get(GLOBAL_SERVER_WORKUNIT)));
     boost::algorithm::ireplace_first(str, _T("/WsWorkunits"), _T("/esp/files/stub.htm?Widget=WUDetailsWidget&Wuid="));
     url = str.c_str();
+    url = OverRideEclWatchURL(url);
```
Comment (GordonSmith) [inline eclide/ResultView.cpp:65]: Instead of this, why not change line 62 to use GLOBAL_ECLWATCH?

File: eclide/ResultView.cpp [hunk 6]
```diff
@@ -71,6 +73,7 @@ const TCHAR * GetGraphEclWatchURL(Dali::IWorkunit *wu, CString &url)
     std::_tstring str = static_cast<const TCHAR * >(CString(GetIConfig(QUERYBUILDER_CFG)->Get(GLOBAL_SERVER_WORKUNIT)));
     boost::algorithm::ireplace_first(str, _T("/WsWorkunits"), _T("/esp/files/stub.htm?Widget=GraphsWUWidget&Wuid="));
     url = str.c_str();
+    url = OverRideEclWatchURL(url);
```
Comment (GordonSmith) [inline eclide/ResultView.cpp:76]: Instead of this, why not change line 73 to use GLOBAL_ECLWATCH?

File: eclide/ResultView.cpp [hunk 7]
```diff
@@ -83,6 +86,7 @@ const TCHAR * GetResultEclWatchURL(Dali::IWorkunit *wu, CString &url, int sequen
     std::_tstring str = static_cast<const TCHAR * >(CString(GetIConfig(QUERYBUILDER_CFG)->Get(GLOBAL_SERVER_WORKUNIT)));
     boost::algorithm::ireplace_first(str, _T("/WsWorkunits"), _T("/esp/files/stub.htm?Widget=ResultWidget&Wuid="));
     url = str.c_str();
+    url = OverRideEclWatchURL(url);
```
Comment (GordonSmith) [inline eclide/ResultView.cpp:89]: Instead of this, why not change line 86 to use GLOBAL_ECLWATCH?

File: eclide/ResultView.cpp [hunk 8]
```diff
@@ -100,13 +104,27 @@ const TCHAR * GetDesdlEclWatchURL(const std::_tstring & desdlID, const std::_tst
     std::_tstring str = static_cast<const TCHAR * >(CString(GetIConfig(QUERYBUILDER_CFG)->Get(GLOBAL_SERVER_WORKUNIT)));
     boost::algorithm::ireplace_first(str, _T("/WsWorkunits"), _T("/esp/files/stub.htm?Widget=DynamicESDLDefinitionDetailsWidget&Id="));
     url = str.c_str();
+    url = OverRideEclWatchURL(url);
```
Comment (GordonSmith) [inline eclide/ResultView.cpp:107]: Instead of this, why not change line 104 to use GLOBAL_ECLWATCH?

**Review by GordonSmith — CHANGES_REQUESTED** [review]

**Comment by dehilsterlexis** [comment]
@GordonSmith I could not find anywhere else GLOBAL_SERVER_WORKUNIT was set so I think we are good. For review.

File: eclide/ResultView.cpp [hunk 9]
```diff
@@ -46,9 +46,10 @@ const TCHAR * GetFramedWorkUnitEclWatchURL(Dali::IWorkunit *wu, CString &url)
 {
     //tp://IP:PORT?inner=/WsWorkunits/WUInfo?Wuid=W20050302-182157
     //http://192.168.1.201:8010/esp/files/stub.htm?Widget=HPCCPlatformWidget#/stub/ECL/Workunits/W20140410-041544/Summary - TODO New EclWatch does not support this fully...
-    std::_tstring str = static_cast<const TCHAR * >(CString(GetIConfig(QUERYBUILDER_CFG)->Get(GLOBAL_SERVER_WORKUNIT)));
+    std::_tstring str = CString(GetIConfig(QUERYBUILDER_CFG)->Get(GLOBAL_ECLWATCH));
     boost::algorithm::ireplace_first(str, _T("/WsWorkunits"), _T("/esp/files/stub.htm?Wuid="));
```
Comment (GordonSmith) [inline eclide/ResultView.cpp:50]: This line is no longer needed.

**Review by GordonSmith — CHANGES_REQUESTED** [review]

File: eclide/ResultView.cpp (hunk 9 above)
Comment (dehilsterlexis) [reply eclide/ResultView.cpp:50]: Will remove

**Comment by dehilsterlexis** [comment]
@GordonSmith for review

**Comment by GordonSmith** [comment]
> @GordonSmith I could not find anywhere else GLOBAL_SERVER_WORKUNIT was set so I think we are good. For review.

For consistency it would be better to add a new item "eclwatchRoot" here: https://github.com/hpcc-systems/eclide/blob/master/comms/comms.cpp#L81-L87 which should have the default text "/esp/files/stub.htm"

**Review by GordonSmith — CHANGES_REQUESTED** [review]
> @GordonSmith I could not find anywhere else GLOBAL_SERVER_WORKUNIT was set so I think we are good. For review.

For consistency it would be better to add a new item "eclwatchRoot" here: https://github.com/hpcc-systems/eclide/blob/master/comms/comms.cpp#L81-L87 which should have the default text "/esp/files/stub.htm"

**Comment by dehilsterlexis** [comment]
@GordonSmith, it seems like we could centralize all these strings so they could be used here as well: https://github.com/hpcc-systems/eclide/blob/84d537be20d26fcd62ece293abfab49bde96602e/wlib/PreferenceDlg.cpp#L454-L460

File: comms/comms.cpp [hunk 10]
```diff
@@ -85,14 +86,15 @@ static const TCHAR accountRoot[]	= _T("8010/Ws_Account");
 static const TCHAR smcRoot[]		= _T("8010/WsSMC");
 static const TCHAR filesprayRoot[]	= _T("8010/FileSpray");
 static const TCHAR dfuRoot[]		= _T("8010/WsDfu");
+static const TCHAR eclwatchRoot[]  = _T("8010/");
```
Comment (GordonSmith) [inline comms/comms.cpp:89]: This should be:  `8010/esp/files/stub.htm`

**Review by GordonSmith — CHANGES_REQUESTED** [review]

File: comms/comms.cpp (hunk 10 above)
Comment (dehilsterlexis) [reply comms/comms.cpp]: Done

**Comment by dehilsterlexis** [comment]
@GordonSmith for review

**Review by GordonSmith — CHANGES_REQUESTED** [review]

File: eclide/ResultView.cpp [hunk 11]
```diff
@@ -46,9 +46,9 @@ const TCHAR * GetFramedWorkUnitEclWatchURL(Dali::IWorkunit *wu, CString &url)
 {
     //tp://IP:PORT?inner=/WsWorkunits/WUInfo?Wuid=W20050302-182157
     //http://192.168.1.201:8010/esp/files/stub.htm?Widget=HPCCPlatformWidget#/stub/ECL/Workunits/W20140410-041544/Summary - TODO New EclWatch does not support this fully...
-    std::_tstring str = static_cast<const TCHAR * >(CString(GetIConfig(QUERYBUILDER_CFG)->Get(GLOBAL_SERVER_WORKUNIT)));
-    boost::algorithm::ireplace_first(str, _T("/WsWorkunits"), _T("/esp/files/stub.htm?Wuid="));
+    std::_tstring str = CString(GetIConfig(QUERYBUILDER_CFG)->Get(GLOBAL_SERVER_ECLWATCH));
     url = str.c_str();
+    url += _T("/esp/files/stub.htm?Wuid=");
```
[the remark this answers is not in the document]
Comment (dehilsterlexis) [reply eclide/ResultView.cpp:51]: @GordonSmith, I'm a bit confused. Wouldn't this generate "esp/files/stub.htm**WORKUNITNUMBER**" with no question mark separator or argument name?

[the remark this answers is not in the document]
Comment (GordonSmith) [reply eclide/ResultView.cpp:51]: As it stands the code will be: `esp/files/stub.htmesp/files/stub.htm?Wuid=WUID` (its `+=` to the global default).

IOW:  `url += _T("?Wuid=");`

**Review by GordonSmith — CHANGES_REQUESTED** [review]
Now the global value has the root url - you only need `url += _T("?Wuid=");` locally?

**Comment by dehilsterlexis** [comment]
@GordonSmith to clarify, the default root ECL Watch URL instead of being (for example) "**http://play.hpccsystems.com:8010/**" should be "**http://play.hpccsystems.com:8010/esp/files/stub.htm**"?

**Comment by GordonSmith** [comment]
The code now looks like this?
```
    std::_tstring str = CString(GetIConfig(QUERYBUILDER_CFG)->Get(GLOBAL_SERVER_ECLWATCH));
    url = str.c_str();
    url += _T("/esp/files/stub.htm?Wuid=");
```
What does `GetIConfig(QUERYBUILDER_CFG)->Get(GLOBAL_SERVER_ECLWATCH)` return?

**Comment by dehilsterlexis** [comment]
@GordonSmith it returns the root only for the default that ends in "...8010", no "esp/files/stub.htm" for now. Do we want to default it to "...8010/esp/files/stub.htm" and let the user change it or do we always need to put "/esp/files/stub.htm"?

**Comment by GordonSmith** [comment]
This is what I see (and expect):
image

image

**Comment by dehilsterlexis** [comment]
@GordonSmith for review

File: eclide/ResultView.cpp [hunk 12]
```diff
@@ -58,19 +58,19 @@ const TCHAR * GetFramedWorkUnitEclWatchURL(Dali::IWorkunit *wu, CString &url)
 const TCHAR * GetWorkUnitEclWatchURL(Dali::IWorkunit *wu, CString &url)
 {
     //http://192.168.1.201:8010/esp/files/stub.htm?Wuid=W20140410-041422&Widget=WUDetailsWidget
-    std::_tstring str = static_cast<const TCHAR * >(CString(GetIConfig(QUERYBUILDER_CFG)->Get(GLOBAL_SERVER_WORKUNIT)));
-    boost::algorithm::ireplace_first(str, _T("/WsWorkunits"), _T("/esp/files/stub.htm?Widget=WUDetailsWidget&Wuid="));
+    std::_tstring str = CString(GetIConfig(QUERYBUILDER_CFG)->Get(GLOBAL_SERVER_ECLWATCH));
     url = str.c_str();
+    url += _T("?Wuid=");
```
Comment (GordonSmith) [inline eclide/ResultView.cpp:63]: This part of the URL should match the bits after the ? in the old URL

File: eclide/ResultView.cpp [hunk 13]
```diff
@@ -58,19 +58,19 @@ const TCHAR * GetFramedWorkUnitEclWatchURL(Dali::IWorkunit *wu, CString &url)
...
     return url;
 }
 
 const TCHAR * GetGraphEclWatchURL(Dali::IWorkunit *wu, CString &url)
 {
-    std::_tstring str = static_cast<const TCHAR * >(CString(GetIConfig(QUERYBUILDER_CFG)->Get(GLOBAL_SERVER_WORKUNIT)));
-    boost::algorithm::ireplace_first(str, _T("/WsWorkunits"), _T("/esp/files/stub.htm?Widget=GraphsWUWidget&Wuid="));
+    std::_tstring str = CString(GetIConfig(QUERYBUILDER_CFG)->Get(GLOBAL_SERVER_ECLWATCH));
     url = str.c_str();
+    url += _T("?Wuid=");
```
Comment (GordonSmith) [inline eclide/ResultView.cpp:73]: This part of the URL should match the bits after the ? in the old URL

File: eclide/ResultView.cpp [hunk 14]
```diff
@@ -80,9 +80,9 @@ const TCHAR * GetResultEclWatchURL(Dali::IWorkunit *wu, CString &url, int sequen
 {
     //http://10.150.64.208:8010/WsWorkunits/WUResult?Wuid=W20041006-101104&Sequence=0
     //http://192.168.1.201:8010/esp/files/stub.htm?Wuid=W20140410-041422&Sequence=0&Widget=ResultWidget
-    std::_tstring str = static_cast<const TCHAR * >(CString(GetIConfig(QUERYBUILDER_CFG)->Get(GLOBAL_SERVER_WORKUNIT)));
-    boost::algorithm::ireplace_first(str, _T("/WsWorkunits"), _T("/esp/files/stub.htm?Widget=ResultWidget&Wuid="));
+    std::_tstring str = CString(GetIConfig(QUERYBUILDER_CFG)->Get(GLOBAL_SERVER_ECLWATCH));
     url = str.c_str();
+    url += _T("?Wuid=");
```
Comment (GordonSmith) [inline eclide/ResultView.cpp:85]: This part of the URL should match the bits after the ? in the old URL

**Review by GordonSmith — CHANGES_REQUESTED** [review]

**Comment by dehilsterlexis** [comment]
@GordonSmith for review

**Review by GordonSmith — APPROVED** [review]

**Review by GordonSmith — APPROVED** [review]

**Pull request merged**